File: plugins/harness/skills/harness-install/scripts/install_harness/preview.py
from __future__ import annotations

import os
from pathlib import Path
import sys

from .errors import fail, read_text
from .models import InstallCandidate, InstallerSupport
from .paths import required_real_target, required_src


class PreviewMixin(InstallerSupport):
    def preview_install_set(self) -> None:

        print(f"target: {Path.cwd()}")
        print(f"mode: {self.config.mode}")
        print(f"ci-host: {self.config.ci_host}")
        print(f"validation command: {self.config.validation_command}")
        for candidate in self.build_plan().candidates:
            self.preview_candidate_status(candidate)

    def preview_candidate_status(self, candidate: InstallCandidate) -> None:

        if candidate.kind in {"file", "seed", "stack-file"}:
            self.preview_file_candidate(candidate)
        elif candidate.kind == "root-contract":
            self.preview_root_contract_candidate(candidate)
        elif candidate.kind == "symlink":
            self.preview_symlink_candidate(candidate)
        elif candidate.kind == "gitkeep":
            self.preview_gitkeep_candidate(candidate)
        else:
            print(f"skip unknown candidate: {candidate.dst}", file=sys.stderr)

    def preview_file_candidate(self, candidate: InstallCandidate) -> None:

        if candidate.seed and Path(candidate.dst).exists():
            print(f"skip seed (target exists): {candidate.dst}")
            return
        if Path(candidate.dst).exists():
            if self.config.force:
                print(f"overwrite (--force): {candidate.dst}")
            else:
                print(f"keep existing: {candidate.dst}")
            return
        print(f"write: {candidate.dst}")

    def preview_root_contract_candidate(self, candidate: InstallCandidate) -> None:

        real_target = required_real_target(candidate)
        if Path(real_target).exists() and candidate.marker in read_text(real_target):
            print(f"skip root contract: {candidate.dst}")
            return
        if Path(real_target).exists() and not self.config.force:
            print(
                f"conflict root contract: {real_target} lacks marker {candidate.marker}; rerun with --force to append",
                file=sys.stderr,
            )
            return
        if Path(real_target).exists():
            print(f"update root contract (--force): {candidate.dst}")
        else:
            print(f"create root contract: {candidate.dst}")

    def preview_symlink_candidate(self, candidate: InstallCandidate) -> None:

        if Path(candidate.dst).is_symlink():
            current_target = os.readlink(candidate.dst)
            if current_target == candidate.symlink_target:
                print(
                    f"skip existing symlink: {candidate.dst} -> {candidate.symlink_target}"
                )
            else:
                print(
                    f"skip existing symlink: {candidate.dst} -> {current_target}",
                    file=sys.stderr,
                )
            return
        if Path(candidate.dst).exists():
            print(f"skip existing: {candidate.dst}")
            return
        print(f"create symlink: {candidate.dst} -> {candidate.symlink_target}")

    def preview_gitkeep_candidate(self, candidate: InstallCandidate) -> None:

        if Path(candidate.dst).exists():
            if self.config.force:
                print(f"overwrite (--force): {candidate.dst}")
            else:
                print(f"keep existing: {candidate.dst}")
            return
        print(f"write: {candidate.dst}")

    def show_one_target_path(self, requested_path: str) -> None:

        candidate = self.build_plan().match(requested_path)
        if candidate.kind in {"file", "seed", "stack-file"}:
            print(self.render_template(required_src(candidate)), end="")
            return
        if candidate.kind == "root-contract":
            real_target = required_real_target(candidate)
            if Path(real_target).exists() and candidate.marker in read_text(
                real_target
            ):
                return
            if Path(real_target).exists():
                print(
                    f"note: requested root contract content would be appended to existing file: {real_target}",
                    file=sys.stderr,
                )
            print(self.render_template(required_src(candidate)), end="")
            return
        if candidate.kind == "gitkeep":
            return
        if candidate.kind == "symlink":
            fail(
                f"--show on symlink entry is not supported; run full install to create {requested_path}"
            )
        fail(f"requested path is not renderable: {requested_path}")

File: plugins/harness/skills/harness-install/scripts/install-harness/preview.ts
import { readlink } from "node:fs/promises";

import { validationCommandForMode } from "./commands.js";
import { renderCandidateContent } from "./content.js";
import { decideFileInstall } from "./decisions.js";
import { isSymlink, matchCandidate, pathExists, readUtf8 } from "./files.js";
import {
  checkSafeFileDestination,
  checkSafeParentDir,
  requiredRealTarget
} from "./paths.js";
import { buildPlan } from "./planning.js";
import { fail } from "./types.js";
import type { InstallCandidate, InstallerConfig } from "./types.js";

const previewFileCandidate = async (
  config: InstallerConfig,
  candidate: InstallCandidate
): Promise<void> => {
  const decision = await decideFileInstall(config, candidate);
  if (decision.diagnostic === "stderr") {
    console.error(decision.message);
  } else {
    console.log(decision.message);
  }
};

const previewRootContractCandidate = async (
  config: InstallerConfig,
  candidate: InstallCandidate
): Promise<void> => {
  const realTarget = requiredRealTarget(candidate);
  await checkSafeFileDestination(realTarget);
  const exists = await pathExists(realTarget);
  if (!exists) {
    console.log(`create root contract: ${candidate.dst}`);
    return;
  }
  const current = await readUtf8(realTarget);
  const source = await renderCandidateContent(candidate);
  if (current === source) {
    console.log(`keep root contract: ${candidate.dst}`);
    return;
  }
  if (config.force) {
    console.log(`update root contract (--force): ${candidate.dst}`);
    return;
  }
  console.error(
    `preserve root contract: ${candidate.dst} differs from the packaged source; rerun with --force to replace it`
  );
};

const previewSymlinkCandidate = async (
  config: InstallerConfig,
  candidate: InstallCandidate
): Promise<void> => {
  await checkSafeParentDir(candidate.dst);
  if (await isSymlink(candidate.dst)) {
    const currentTarget = await readlink(candidate.dst);
    if (currentTarget === candidate.symlinkTarget) {
      console.log(
        `keep existing symlink: ${candidate.dst} -> ${candidate.symlinkTarget ?? ""}`
      );
      return;
    }
    if (config.force) {
      console.log(
        `replace symlink (--force): ${candidate.dst} -> ${candidate.symlinkTarget ?? ""}`
      );
      return;
    }
    console.error(
      `conflict symlink: ${candidate.dst} points to ${currentTarget}; rerun with --force to replace it`
    );
    return;
  }
  if (await pathExists(candidate.dst)) {
    console.error(`conflict symlink: ${candidate.dst} already exists`);
    return;
  }
  console.log(
    `create symlink: ${candidate.dst} -> ${candidate.symlinkTarget ?? ""}`
  );
};

const previewGitkeepCandidate = async (
  config: InstallerConfig,
  candidate: InstallCandidate
): Promise<void> => {
  await checkSafeFileDestination(candidate.dst);
  if (!(await pathExists(candidate.dst))) {
    console.log(`write: ${candidate.dst}`);
    return;
  }
  const current = await readUtf8(candidate.dst);
  if (current === "") {
    console.log(`keep existing: ${candidate.dst}`);
    return;
  }
  if (config.force) {
    console.log(`overwrite (--force): ${candidate.dst}`);
    return;
  }
  console.error(
    `conflict: ${candidate.dst} differs from the packaged empty file; preserving target; rerun with --force to overwrite`
  );
};

const showRootContractCandidate = async (
  candidate: InstallCandidate
): Promise<void> => {
  const realTarget = requiredRealTarget(candidate);
  await checkSafeFileDestination(realTarget);
  process.stdout.write(await renderCandidateContent(candidate));
};

const previewCandidateStatus = async (
  config: InstallerConfig,
  candidate: InstallCandidate
): Promise<void> => {
  switch (candidate.kind) {
    case "file":
    case "seed":
    case "stack-file": {
      await previewFileCandidate(config, candidate);
      return;
    }
    case "root-contract": {
      await previewRootContractCandidate(config, candidate);
      return;
    }
    case "symlink": {
      await previewSymlinkCandidate(config, candidate);
      return;
    }
    case "gitkeep": {
      await previewGitkeepCandidate(config, candidate);
      return;
    }
    default: {
      return fail(`skip unknown candidate: ${candidate.dst}`);
    }
  }
};

const previewCandidates = async (
  config: InstallerConfig,
  candidates: readonly InstallCandidate[],
  index = 0
): Promise<void> => {
  if (index >= candidates.length) {
    return;
  }
  const candidate = candidates[index];
  if (candidate !== undefined) {
    await previewCandidateStatus(config, candidate);
  }
  await previewCandidates(config, candidates, index + 1);
};

/** Print the selected install set and each candidate status without writing files. */
export const previewInstallSet = async (
  config: InstallerConfig
): Promise<void> => {
  console.log(`target: ${process.cwd()}`);
  console.log(`mode: ${config.mode}`);
  console.log(`ci-host: ${config.ciHost}`);
  console.log(`validation command: ${validationCommandForMode(config.mode)}`);
  await previewCandidates(config, await buildPlan(config));
};

/** Print rendered content for one target path without writing files. */
export const showOneTargetPath = async (
  _config: InstallerConfig,
  requestedPath: string
): Promise<void> => {
  const candidate = matchCandidate(await buildPlan(_config), requestedPath);
  switch (candidate.kind) {
    case "file":
    case "seed":
    case "stack-file": {
      await checkSafeFileDestination(candidate.dst);
      process.stdout.write(await renderCandidateContent(candidate));
      return;
    }
    case "root-contract": {
      await showRootContractCandidate(candidate);
      return;
    }
    case "gitkeep": {
      await checkSafeFileDestination(candidate.dst);
      return;
    }
    case "symlink": {
      return fail(
        `--show on symlink entry is not supported; run full install to create ${requestedPath}`
      );
    }
    default: {
      return fail(`requested path is not renderable: ${requestedPath}`);
    }
  }
};
